### PR TITLE
Meeting Component

### PR DESCRIPTION
**Pull request by akoushke**

### CODE REVIEW

File: src/components/WebexInMeeting/README.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# Webex Interstitial Meeting Component
```
Comment (adamweeks) [inline src/components/WebexInMeeting/README.md]:
```suggestion
# Webex InMeeting Component
```

File: src/components/WebexInMeeting/README.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+Webex InMeeting component displays the [Webex Local Media](../WebexLocalMedia) component on top of the [Webex Remote Media](../WebexRemoteMedia) component for the in meeting media experience.
+
+<p align="center">
+  <img src="./WebexInMeeting.png" alt="Default Webex InMeeting" />
+</p>
+
+## Preview
+
+To see all the different possible states of the Webex Interstitial Meeting component, you can run our Storybook:
```
Comment (adamweeks) [inline src/components/WebexInMeeting/README.md]:
```suggestion
To see all the different possible states of the Webex InMeeting component, you can run our Storybook:
```

**Comment by akoushke** [comment]
will clean up the commits after review.

File: src/components/hooks/useMeeting.js [hunk 3]
```diff
@@ -22,9 +22,12 @@ export default function useMeeting(meetingID, meetingDestination) {
     const onError = (error) => {
       throw error;
     };
+    const onMeeting = (newMeeting) => {
```
Comment (lalli-flores) [inline src/components/hooks/useMeeting.js]: Can you add a comment as to why we're doing this?
```suggestion
    // React won't recognize the meeting attributes have been updated
    // since the state is the meeting object itself. We need to create a new
    // meeting object trigger the state change
    const onMeeting = (newMeeting) => {
```

File: src/components/hooks/useMeeting.js [hunk 4]
```diff
@@ -33,9 +36,10 @@ export default function useMeeting(meetingID, meetingDestination) {
         .pipe(
           flatMap(({ID}) => meetingsAdapter.getMeeting(ID)),
           take(1),
-          delayWhen(({ID}) => from(meetingsAdapter.addLocalMedia(ID)))
+          delayWhen(({ID}) => from(meetingsAdapter.addLocalMedia(ID))),
+          flatMap(({ID}) => meetingsAdapter.getMeeting(ID))
```
Comment (lalli-flores) [inline src/components/hooks/useMeeting.js]: Why did you move this after adding local media? I thought the idea of needing this here was because the event subscription needs to happen before the event (adding local media)

File: src/components/WebexMeeting/WebexMeeting.test.js [hunk 5]
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
+
+import WebexInterstitialMeeting from './WebexMeeting';
```
Comment (lalli-flores) [inline src/components/WebexMeeting/WebexMeeting.test.js]:
```suggestion
import WebexMeeting from './WebexMeeting';
```

File: src/components/WebexMeeting/WebexMeeting.test.js [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+jest.mock('../hooks/useMeeting');
+
+describe('Webex Interstitial Meeting component', () => {
```
Comment (lalli-flores) [inline src/components/WebexMeeting/WebexMeeting.test.js]:
```suggestion
describe('Webex Meeting component', () => {
```

File: src/components/WebexMeeting/WebexMeeting.test.js [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+  describe('snapshot', () => {
+    test('matches snapshot of loading while meeting is created', () => {
+      expect(shallow(<WebexInterstitialMeeting meetingDestination="" />)).toMatchSnapshot();
+    });
+
+    test('matches snapshot of meeting destination', () => {
+      expect(shallow(<WebexInterstitialMeeting meetingDestination="localMedia" />)).toMatchSnapshot();
+    });
+  });
+});
```
Comment (lalli-flores) [inline src/components/WebexMeeting/WebexMeeting.test.js]: There's a snapshot missing for the "in-meeting" experience

File: src/components/WebexMeeting/WebexMeeting.test.js [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+jest.mock('../hooks/useMeeting');
+
+describe('Webex Interstitial Meeting component', () => {
+  describe('snapshot', () => {
+    test('matches snapshot of loading while meeting is created', () => {
+      expect(shallow(<WebexInterstitialMeeting meetingDestination="" />)).toMatchSnapshot();
+    });
+
+    test('matches snapshot of meeting destination', () => {
```
Comment (lalli-flores) [inline src/components/WebexMeeting/WebexMeeting.test.js]:
```suggestion
    test('matches snapshot of a user waiting to join a meeting', () => {
```

File: src/components/WebexMeeting/WebexMeeting.stories.js [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+import {WebexJSONAdapter} from '../../adapters';
+
+import {WebexMeeting, WebexDataProvider} from '..';
+
+// Setup for the stories
+const stories = storiesOf('Webex Meeting', module);
+const webexAdapter = new WebexJSONAdapter(jsonData);
+const wrapperStyle = {height: '500px', width: '800px', border: '1px solid black'};
+
+stories.add('in session', () => (
```
Comment (lalli-flores) [inline src/components/WebexMeeting/WebexMeeting.stories.js]: What does this story represent?

File: src/components/WebexMeeting/README.md [hunk 10]
```diff
@@ -0,0 +1,35 @@
+# Webex Interstitial Meeting Component
+
+Webex Interstitial meeting component displays the [Webex Local Media](../WebexLocalMedia) and [Webex Meeting Info](../WebexMeetingInfo) components for a meeting.
+
+<p align="center">
+  <img src="./WebexInterstitialMeeting.png" alt="Default Webex Interstitial Meeting" />
```
Comment (lalli-flores) [inline src/components/WebexMeeting/README.md]: This picture needs to be a gif, showing the interstitial => in-meeting experience.

File: src/components/WebexMeeting/README.md [hunk 11]
```diff
@@ -0,0 +1,35 @@
+# Webex Interstitial Meeting Component
```
Comment (lalli-flores) [inline src/components/WebexMeeting/README.md]:
```suggestion
# Webex Meeting Component
```

File: src/components/WebexMeeting/README.md [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+
+Webex Interstitial meeting component displays the [Webex Local Media](../WebexLocalMedia) and [Webex Meeting Info](../WebexMeetingInfo) components for a meeting.
```
Comment (lalli-flores) [inline src/components/WebexMeeting/README.md]:
```suggestion
Webex Meeting component displays the complete, default Webex meeting experience.
```

File: src/components/WebexMeeting/README.md [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+
+<p align="center">
+  <img src="./WebexInterstitialMeeting.png" alt="Default Webex Interstitial Meeting" />
+</p>
+
+## Preview
+
+To see all the different possible states of the Webex Interstitial Meeting component, you can run our Storybook:
```
Comment (lalli-flores) [inline src/components/WebexMeeting/README.md]:
```suggestion
To see all the different possible states of the Webex Meeting component, you can run our Storybook:
```

File: src/components/WebexMeeting/README.md [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+
+## Embed
+
+1.  Create a component adapter from which the data will be retrieved (See [adapters](../../adapters)). For instance:
+
+    ```js
+    const jsonAdapter = new WebexJSONAdapter(jsonData);
+    ```
+
+2.  Create a component instance by passing the person ID as a string and
```
Comment (lalli-flores) [inline src/components/WebexMeeting/README.md]:
```suggestion
2.  Create a component instance by passing the meeting ID as a string and
```

File: src/components/WebexMeeting/README.md [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+    enclose it within [a data provider](../WebexDataProvider/WebexDataProvider.js)
+    that takes the [component data adapter](../../adapters/WebexJSONAdapter.js) that we created previously
+
+    ```js
+    <WebexDataProvider adapter={jsonAdapter}>
+      <WebexInterstitialMeeting meetingID="meetingID" />
```
Comment (lalli-flores) [inline src/components/WebexMeeting/README.md]: Can you fix this to match the actual component?

File: src/components/WebexMeeting/WebexMeeting.js [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+WebexMeeting.propTypes = {
+  meetingDestination: PropTypes.string.isRequired,
+  setControls: PropTypes.func,
+};
+
+WebexMeeting.defaultProps = {
+  // eslint-disable-next-line no-confusing-arrow
+  setControls: (isActive) =>
```
Comment (lalli-flores) [inline src/components/WebexMeeting/WebexMeeting.js]: Please add a `func` typedef explaining this function prop

File: src/components/WebexMeeting/WebexMeeting.js [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+      {isActive ? <WebexInMeeting meetingID={ID} /> : <WebexInterstitialMeeting meetingID={ID} />}
+      <div className="controls">
+        <WebexMeetingControls meetingID={ID}>{controls}</WebexMeetingControls>
+      </div>
+    </Fragment>
+  ) : (
+    <Spinner />
+  );
+
+  return <div className="meeting">{meeting}</div>;
```
Comment (lalli-flores) [inline src/components/WebexMeeting/WebexMeeting.js]: Can you move the `meeting` variable logic here? Not sure why it's separated out. It would also make the `Fragment` unnecessary

File: src/components/WebexMeeting/WebexMeeting.js [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+export default function WebexMeeting({meetingDestination, setControls}) {
+  const {ID, remoteVideo} = useMeeting(null, meetingDestination);
+
+  const isActive = remoteVideo !== null;
+
+  const controls = setControls(isActive).map((key) => <WebexMeetingControl type={key} key={key} />);
+  const meeting = ID ? (
+    <Fragment>
+      {isActive ? <WebexInMeeting meetingID={ID} /> : <WebexInterstitialMeeting meetingID={ID} />}
+      <div className="controls">
```
Comment (lalli-flores) [inline src/components/WebexMeeting/WebexMeeting.js]: Does `WebexMeetingControls` create a `div` itself? if not, I think this should go within that component, since `WebexMeetingControls` will always wrap `WebexMeetingControl`

File: src/components/WebexMeeting/WebexMeeting.js [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+ *
+ * @param {object} props
+ * @returns {object} JSX of the component
+ */
+export default function WebexMeeting({meetingDestination, setControls}) {
+  const {ID, remoteVideo} = useMeeting(null, meetingDestination);
+
+  const isActive = remoteVideo !== null;
+
+  const controls = setControls(isActive).map((key) => <WebexMeetingControl type={key} key={key} />);
```
Comment (lalli-flores) [inline src/components/WebexMeeting/WebexMeeting.js]:
```suggestion
  const controls = setControls(isActive).map((key) => <WebexMeetingControl key={key} type={key} />);
```

File: src/components/WebexMeeting/WebexMeeting.css [hunk 20]
```diff
@@ -0,0 +1,15 @@
...
+  align-items: center;
+  height: 100%;
+  width: 100%;
+}
+
+.controls {
+  display: flex;
+  position: absolute;
+  align-self: flex-end;
+  padding-bottom: 3rem;
```
Comment (lalli-flores) [inline src/components/WebexMeeting/WebexMeeting.css]: If you're going to absolute position it, use the positioning properties instead of padding

**Comment by akoushke** [comment]
> I didn't realize you were going to "implement" the controls in the JSON adapter. Is that necessary?

whatcha mean? we had `audio` control implemented already. I just added the other controls

File: src/components/hooks/useMeeting.js (hunk 4 above)
Comment (akoushke) [reply src/components/hooks/useMeeting.js]: so I couldn't come up with a better approach than resubscribing to the meeting after addLocalMedia. basically what I need to do is to trigger `addLocalMedia` only once after `getMeeting` subscription.

Comment (akoushke) [reply src/components/hooks/useMeeting.js]: i didn't move it. I appended another `getMeeting`

File: src/components/WebexMeeting/WebexMeeting.stories.js (hunk 9 above)
Comment (akoushke) [reply src/components/WebexMeeting/WebexMeeting.stories.js]: it's the whole experience

File: src/components/WebexMeeting/WebexMeeting.js (hunk 16 above)
Comment (akoushke) [reply src/components/WebexMeeting/WebexMeeting.js]: not sure how you want it.

File: src/components/WebexMeeting/WebexMeeting.js (hunk 17 above)
Comment (akoushke) [reply src/components/WebexMeeting/WebexMeeting.js]: we still need the `Fragment`. but I did move it in.

File: src/components/hooks/useMeetingControl.js [hunk 21]
```diff
@@ -12,9 +12,15 @@ export default function useMeetingControl(type) {
...
+  const controls = meetingsAdapter.meetingControls;
+  let control;
 
-  const control = meetingsAdapter.meetingControls[type];
-  const {action} = control;
+  if (controls[type]) {
+    control = controls[type];
+  } else {
+    // eslint-disable-next-line no-console
+    console.error(`${type} control is not defined. Available controls are "${Object.keys(controls).join(', ')}".`);
```
Comment (lalli-flores) [inline src/components/hooks/useMeetingControl.js]: I like that we let the developer know what controls are implemented 👍🏻

File: src/components/WebexMeeting/WebexMeeting.test.js [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+  describe('snapshot', () => {
+    test('matches snapshot of loading while meeting is created', () => {
+      expect(shallow(<WebexMeeting meetingDestination="" />)).toMatchSnapshot();
+    });
+
+    test('matches snapshot of a user waiting to join a meeting', () => {
+      expect(shallow(<WebexMeeting meetingDestination="localMedia" />)).toMatchSnapshot();
+    });
+
+    test('matches snapshot of a user had joined a meeting', () => {
```
Comment (lalli-flores) [inline src/components/WebexMeeting/WebexMeeting.test.js]:
```suggestion
    test('matches snapshot of a user that has joined a meeting', () => {
```

File: src/components/hooks/useMeeting.js (hunk 4 above)
Comment (lalli-flores) [reply src/components/hooks/useMeeting.js]: We can address this outside of the PR, but I something tells me this is not the right thing to do

File: src/adapters/MeetingsJSONAdapter.js [hunk 23]
```diff
@@ -143,13 +157,25 @@ export default class MeetingsJSONAdapter extends MeetingsAdapter {
...
+    const leaveEvents$ = fromEvent(document, LEAVE_CONTROL);
+
+    const events$ = merge(audioEvents$, videoEvents$, joinEvents$, leaveEvents$).pipe(
       filter((event) => event.detail.ID === ID),
       map((event) => event.detail)
     );
 
-    return concat(getMeetingWithMedia$, muteEvent$);
+    return concat(getMeetingWithMedia$, events$).pipe(
+      map((meeting) => {
```
Comment (lalli-flores) [inline src/adapters/MeetingsJSONAdapter.js]: This could be a `tap`
```suggestion
      tap((meeting) => this.datasource[ID] = meeting)
```

File: src/adapters/MeetingsJSONAdapter.js [hunk 24]
```diff
@@ -201,6 +227,88 @@ export default class MeetingsJSONAdapter extends MeetingsAdapter {
...
+
+      meeting.remoteVideo = await this.getStream({video: true, audio: false});
+      meeting.remoteAudio = await this.getStream({video: false, audio: true});
+
+      document.dispatchEvent(new CustomEvent(JOIN_CONTROL, {detail: meeting}));
+    }
+  }
+
+  /**
+   * Join the meeting by nullifying the remote media
```
Comment (lalli-flores) [inline src/adapters/MeetingsJSONAdapter.js]:
```suggestion
   * Join the meeting by removing the remote media
```

File: src/adapters/MeetingsJSONAdapter.js [hunk 25]
```diff
@@ -201,6 +227,88 @@ export default class MeetingsJSONAdapter extends MeetingsAdapter {
...
+   * Toggles muting the local audio media stream track.
+   * Used by "mute-video" meeting control.
+   *
+   * @param {string} ID  ID of the meeting for which to mute local audio.
+   * @memberof MeetingsJSONAdapter
+   * @private
+   */
+  async toggleMuteVideo(ID) {
+    if (this.datasource[ID]) {
+      const meeting = this.datasource[ID];
```
Comment (lalli-flores) [inline src/adapters/MeetingsJSONAdapter.js]: This is why the rending bug was showing up. If you look at the previous implementation, a new object was always created and returned. Here we were sending back the same object

File: src/components/WebexMeeting/README.md [hunk 26]
```diff
@@ -0,0 +1,37 @@
...
+    ```js
+    const jsonAdapter = new WebexJSONAdapter(jsonData);
+    ```
+
+2.  Create a component instance by passing the meeting ID as a string and
+    enclose it within [a data provider](../WebexDataProvider/WebexDataProvider.js)
+    that takes the [component data adapter](../../adapters/WebexJSONAdapter.js) that we created previously
+
+    ```js
+    const setControls = (isActive) => isActive ? ['join-meeting'] : ['leave-meeting'];
```
[the remark this answers is not in the document]
Comment (akoushke) [reply src/components/WebexMeeting/README.md]: I added more description

File: src/adapters/MeetingsJSONAdapter.js (hunk 25 above)
Comment (akoushke) [reply src/adapters/MeetingsJSONAdapter.js]: right. the static data needs to be updated though.

Comment (lalli-flores) [reply src/adapters/MeetingsJSONAdapter.js]: We should a copy of the meeting before modifying it, to treat it as "immutable"
```suggestion
      const meeting = {...this.datasource[ID]};
```

I know we fixed it in the hook, but I think it's good to show "proper" code

File: src/adapters/MeetingsJSONAdapter.js [hunk 27]
```diff
@@ -201,6 +227,88 @@ export default class MeetingsJSONAdapter extends MeetingsAdapter {
...
+   * Toggles muting the local audio media stream track.
+   * Used by "mute-audio" meeting control.
+   *
+   * @param {string} ID  ID of the meeting for which to mute local audio.
+   * @memberof MeetingsJSONAdapter
+   * @private
+   */
+  async toggleMuteAudio(ID) {
+    if (this.datasource[ID]) {
+      const meeting = this.datasource[ID];
```
Comment (lalli-flores) [inline src/adapters/MeetingsJSONAdapter.js]: I know we fixed it in the hook, but I think it's good to show "proper" code
```suggestion
      const meeting = {...this.datasource[ID]};
```

File: src/adapters/MeetingsJSONAdapter.js [hunk 28]
```diff
@@ -143,13 +157,25 @@ export default class MeetingsJSONAdapter extends MeetingsAdapter {
...
-    const muteEvent$ = fromEvent(document, MUTE_AUDIO_CONTROL).pipe(
+    // Send updates on the meeting when an action is triggered
+    const audioEvents$ = fromEvent(document, MUTE_AUDIO_CONTROL);
+    const videoEvents$ = fromEvent(document, MUTE_VIDEO_CONTROL);
+    const joinEvents$ = fromEvent(document, JOIN_CONTROL);
+    const leaveEvents$ = fromEvent(document, LEAVE_CONTROL);
+
+    const events$ = merge(audioEvents$, videoEvents$, joinEvents$, leaveEvents$).pipe(
       filter((event) => event.detail.ID === ID),
       map((event) => event.detail)
```
Comment (lalli-flores) [inline src/adapters/MeetingsJSONAdapter.js]:
```suggestion
      map((event) => {...event.detail}) // Make a copy of the meeting to treat it as if were immutable
```

File: src/adapters/MeetingsJSONAdapter.js (hunk 25 above)
Comment (lalli-flores) [reply src/adapters/MeetingsJSONAdapter.js]: Actually, if we return a new object in `getMeeting` all the time, it should work there. Left a comment around that code instead

File: src/adapters/MeetingsJSONAdapter.test.js [hunk 29]
```diff
@@ -79,14 +80,50 @@ describe('Meetings JSON Adapter', () => {
...
-    test('returns meeting data from mute event', (done) => {
+    test('returns meeting data from mute audio event', (done) => {
       meetingID = 'localAudio';
 
       meetingsJSONAdapter
         .getMeeting(meetingID)
         .pipe(tap(() => meetingsJSONAdapter.toggleMuteAudio(meetingID)))
         .subscribe((meeting) => {
-          expect(meeting.localAudio).toEqual(['remote-audio']);
+          expect(meeting.localAudio).toEqual(null);
```
Comment (lalli-flores) [inline src/adapters/MeetingsJSONAdapter.test.js]:
```suggestion
          expect(meeting.localAudio).toBeNull();
```

File: src/adapters/MeetingsJSONAdapter.test.js [hunk 30]
```diff
@@ -79,14 +80,50 @@ describe('Meetings JSON Adapter', () => {
...
+    });
+
+    test('returns meeting data from mute video event', (done) => {
+      meetingID = 'localVideo';
+
+      meetingsJSONAdapter
+        .getMeeting(meetingID)
+        .pipe(tap(() => meetingsJSONAdapter.toggleMuteVideo(meetingID)))
+        .subscribe((meeting) => {
+          expect(meeting.localVideo).toEqual(null);
```
Comment (lalli-flores) [inline src/adapters/MeetingsJSONAdapter.test.js]:
```suggestion
          expect(meeting.localVideo).toBeNull();
```

File: src/adapters/MeetingsJSONAdapter.test.js [hunk 31]
```diff
@@ -79,14 +80,50 @@ describe('Meetings JSON Adapter', () => {
...
+    });
+
+    test('returns meeting data from leave event', (done) => {
+      meetingID = 'remoteMedia';
+
+      meetingsJSONAdapter
+        .getMeeting(meetingID)
+        .pipe(tap(() => meetingsJSONAdapter.leaveMeeting(meetingID)))
+        .subscribe((meeting) => {
+          expect(meeting.remoteVideo).toEqual(null);
```
Comment (lalli-flores) [inline src/adapters/MeetingsJSONAdapter.test.js]:
```suggestion
          expect(meeting.remoteVideo).toBeNull();
```

**Comment by lalli-flores** [comment]
Just breakdown the commits and it looks good to go
